### PR TITLE
Save as draft

### PR DESCRIPTION
**Pull request by loneil**
# Description
https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-1492
Save button and edit draft buttons in Submitter view submissions section.
New flag on the forms setting to choose whether form submissions can allow drafts.
Does not apply to public forms. Draft history will not show in edit history for Staff.

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-184

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-184

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-184

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-184

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-184

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-184

File: app/frontend/src/components/designer/FormViewerActions.vue [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      default: false,
+    },
+    submissionId: {
+      type: String,
+      default: undefined,
+    },
+  },
+  data() {
+    return {
+      dialog: false,
```
Comment (jujaga) [inline app/frontend/src/components/designer/FormViewerActions.vue:81]: Is the `dialog` element being used at all in the component? If not, we could completely remove the entire data option from this component.

File: app/frontend/src/components/forms/FormSubmission.vue [hunk 2]
```diff
@@ -175,6 +176,14 @@ export default {
   computed: mapGetters('form', ['form', 'formSubmission', 'permissions']),
   methods: {
     ...mapActions('form', ['fetchSubmission', 'getFormPermissionsForUser']),
+    deleted() {
```
Comment (jujaga) [inline app/frontend/src/components/forms/FormSubmission.vue:179]: Optional: May be better to change this to be something like "onDelete" to better reflect that this is in response to an action.

File: app/frontend/src/components/forms/submission/MySubmissionsTable.vue [hunk 3]
```diff
@@ -4,7 +4,8 @@
       <v-row class="mt-6" no-gutters>
         <v-col class="text-center" cols="12" sm="10" offset-sm="1">
           <h1>Previous Submissions</h1>
-          <h2>{{ form.name }}</h2>
+          <h2 v-if="formId">{{ form.name }}</h2>
+          <h2 v-else>All Forms</h2>
```
Comment (jujaga) [inline app/frontend/src/components/forms/submission/MySubmissionsTable.vue:8]: Should this be in the other format?
```html
<h2>{{ formId ? form.name : 'All Forms' }}</h2>
```

File: app/frontend/src/components/forms/submission/MySubmissionsTable.vue [hunk 4]
```diff
@@ -150,31 +144,42 @@ export default {
...
+      }
     },
-    getStatusDate(history, statusCode) {
+    getStatusDate(record, statusCode) {
       // Get the created date of the most recent occurence of a specified status
-      if (history) {
-        const submittedStatus = history.find((stat) => stat.code == statusCode);
+      if (record.submissionStatus) {
+        const submittedStatus = record.submissionStatus.find(
+          (stat) => stat.code == statusCode
```
Comment (jujaga) [inline app/frontend/src/components/forms/submission/MySubmissionsTable.vue:161]: Triple equals

File: app/frontend/src/views/user/SubmissionDraftEdit.vue [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<template>
+  <div>
+    <UserSubmission :saved="sv" :submissionId="s" :readOnly="false" />
+  </div>
```
Comment (jujaga) [inline app/frontend/src/views/user/SubmissionDraftEdit.vue:4]: Don't need the wrapper `<div>` elements.

File: app/src/db/migrations/20210614134214_017-alter-submission-submitters-add-form.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  INNER JOIN form_version fv ON
+    fv."id" = sub."formVersionId"
+  INNER JOIN form f ON
+    fv."formId" = f."id"`));
+
+};
+
+exports.down = function (knex) {
+  return Promise.resolve()
+
```
Comment (jujaga) [inline app/src/db/migrations/20210614134214_017-alter-submission-submitters-add-form.js:39]: Extra whitespace lines on Lines 34 and 39.

File: app/src/forms/submission/service.js [hunk 7]
```diff
@@ -42,7 +43,26 @@ const service = {
...
+          await service._createSubmissionStatus(formSubmissionId, { code: Statuses.SUBMITTED }, currentUser);
+        }
+
+        // If the state is finalized to submitted, then remove permissions that would allow an edit or delete for submitters
+        // Danger be here, do not mess up the where clauses
+        // (always always always ensure submission id is enforced and you know what KNEX is doing about chaining the where clauses as to if it's making an AND or an OR)
+        await FormSubmissionUser.query(trx).delete()
+          .where('formSubmissionId', formSubmissionId)
+          .whereIn('permission', [Permissions.SUBMISSION_DELETE, Permissions.SUBMISSION_UPDATE]);
+
```
Comment (jujaga) [inline app/src/forms/submission/service.js:64]: Extra whitespace line

**Review by jujaga — APPROVED** [review]
Looks great. Tentatively approving with minor code fixes.

**Comment by TimCsaky** [comment]
looks awesome. People are going to love this!
only thing i thought i should mention, and i expect it is intentional.. and i havent even found the code..
in UI, i found the form owner can view a submitters draft, if they know the URL. i suppose that could be a good thing.
I couldnt get more than one test idir to work (csstrol3) to test this further

**Review by TimCsaky — APPROVED** [review]

**Review by MichaelChung123 — APPROVED** [review]

**Comment by MichaelChung123** [comment]
Approved. Let's try to satisfy the codeclimate warnings before merging, otherwise looks like a great feature!

**Comment by loneil** [comment]
> looks awesome. People are going to love this!
> only thing i thought i should mention, and i expect it is intentional.. and i havent even found the code..
> in UI, i found the form owner can view a submitters draft, if they know the URL. i suppose that could be a good thing.
> I couldnt get more than one test idir to work (csstrol3) to test this further

Yeah eventually viewing a draft and/or details about that draft could be an operational task so we'd want to keep that open for the possibility I think. But just hiding it away from the staff UI for the time being.

File: app/frontend/src/components/forms/submission/MySubmissionsTable.vue (hunk 3 above)
Comment (loneil) [reply app/frontend/src/components/forms/submission/MySubmissionsTable.vue:8]: I don't know, stylistically I kind of like the v-if dom element style rather then ternaries in expressions. We probably mix the 2 all over the place. what do you think is cleaner?

File: app/frontend/src/components/designer/FormViewerActions.vue (hunk 1 above)
Comment (loneil) [reply app/frontend/src/components/designer/FormViewerActions.vue:81]: Fixed

File: app/frontend/src/components/forms/FormSubmission.vue (hunk 2 above)
Comment (loneil) [reply app/frontend/src/components/forms/FormSubmission.vue:179]: Yeah that's a good idea. Using `onX` in emitted event handlers by the parent would be a good pattern

File: app/frontend/src/components/forms/submission/MySubmissionsTable.vue (hunk 4 above)
Comment (loneil) [reply app/frontend/src/components/forms/submission/MySubmissionsTable.vue:161]: Fixed

File: app/frontend/src/views/user/SubmissionDraftEdit.vue (hunk 5 above)
Comment (loneil) [reply app/frontend/src/views/user/SubmissionDraftEdit.vue:4]: Fixed

File: app/src/db/migrations/20210614134214_017-alter-submission-submitters-add-form.js (hunk 6 above)
Comment (loneil) [reply app/src/db/migrations/20210614134214_017-alter-submission-submitters-add-form.js:39]: Fixed

File: app/src/forms/submission/service.js (hunk 7 above)
Comment (loneil) [reply app/src/forms/submission/service.js:64]: Fixed

File: app/frontend/src/components/forms/submission/MySubmissionsTable.vue (hunk 3 above)
Comment (loneil) [reply app/frontend/src/components/forms/submission/MySubmissionsTable.vue:8]: Fixed

**Pull request merged**